### PR TITLE
Single DB

### PR DESCRIPTION
**Pull request by tomusdrw**
1. Consolidating Blocks, State, Extras and Traces DBs into single one using RocksDB Column Families
2. DB Migration

I've done the easier part, most important bits done by @arkpar :)

### CODE REVIEW

**Comment by tomusdrw** [comment]
Would be nice to do some more testing (especially migration from 6 -> 8).

I've tested:
1. Sync from 0
2. Migration from 7->8 & sync

**Comment by gavofyork** [comment]
wow. that was quick!

generally looks good after a 10 minute review. 

would perhaps be nice to avoid the double-barrel arguments of `db, column` which have been introduced and just have an abstraction that covers both and provides what `db` used to before this PR?

can be a later refactoring PR, though.

i'll sync from 0 tonight/tomorrow. i might be able to do the migration from 6->8, too.

**Comment by arkpar** [comment]
I did a full sync. Speed is about the same, but I ended up with a database that has a few thousand files. Something is wrong with passing compaction options.

**Comment by gavofyork** [comment]
i fear the compression merge made a few conflicts...

**Comment by keorn** [comment]
its mostly migrations, anyway to get 6->8 working #1747 is needed.

**Comment by keorn** [comment]
Im merging and testing 6->9.

**Comment by tomusdrw** [comment]
It might be difficult to provide similar interface, cause right now `Database` doesn't provide "write" traits. You need to create multi-column transaction (batch) and write the whole batch.

I think this is the whole point of having atomic writes in multiple columns, providing something similar to previous interface would make it much easier to do single-column transactions (which might not really be  desired).

**Comment by arkpar** [comment]
I was also thinking of introducing `SubBatch` and `SubDatabase` abstractions that would provide a view into a single column, but suggest leaving it to another PR

File: ethcore/src/client/client.rs [hunk 1]
```diff
@@ -141,6 +142,19 @@ pub struct Client {
 }
 
 const HISTORY: u64 = 1200;
+// database columns
+/// Meta(?)
+pub const DB_COL_META: Option<u32> = Some(0);
```
Comment (arkpar) [inline ethcore/src/client/client.rs]: I've added this for meta-DB values, such as version. Can be removed if it is not used.

File: util/src/kvdb.rs [hunk 2]
```diff
@@ -112,7 +112,7 @@ impl Default for DatabaseConfig {
 	fn default() -> DatabaseConfig {
 		DatabaseConfig {
 			cache_size: None,
-			max_open_files: 256,
+			max_open_files: 1024,
```
Comment (debris) [inline util/src/kvdb.rs]: mac os won't handle that much

Comment (tomusdrw) [inline util/src/kvdb.rs]: We already had 4 databases with 256 limit, don't we bump iolimit too?

Comment (debris) [inline util/src/kvdb.rs]: we can try...

**Comment by petevine** [comment]
I believe `util` hasn't been updated:

```shell
Compiling rocksdb v0.4.5 (https://github.com/ethcore/rust-rocksdb#dd597245)
src/kvdb.rs:21:67: 21:73 error: unresolved import `rocksdb::Column`. There is no `Column` in `rocksdb` [E0432]
src/kvdb.rs:21  Options, DBCompactionStyle, BlockBasedOptions, Direction, Cache, Column};
                                                                                 ^~~~~~
src/kvdb.rs:21:67: 21:73 help: run `rustc --explain E0432` to see a detailed explanation
src/kvdb.rs:21:67: 21:73 error: unresolved import `rocksdb::Column`. There is no `Column` in `rocksdb` [E0432]
src/kvdb.rs:21  Options, DBCompactionStyle, BlockBasedOptions, Direction, Cache, Column};
                                                                                 ^~~~~~
````

**Comment by arkpar** [comment]
there's no Cargo.lock file in the sub-crates. You'll have to run `cargo update -p rocksdb` manually if you want to build util crate separately.
Alternatively you can run `cargo build -p ethcore-util` from the root

**Comment by rphmeier** [comment]
@petevine no, util is updated -- it wouldn't build in travis if it wasn't, and my local builds are all successful as well. probably a cargo issue. please try the following:
  - ensure there isn't a Cargo.lock in the `util` directory.
  - `cargo clean -p ethcore-util rust-rocksdb`

**Comment by petevine** [comment]
I was running `cargo bench` in there and can't remember having to update the dependencies manually in the past.

**Pull request merged**